tbot_contrib/locking.py

```diff
@@ -0,0 +1,302 @@
+""" Machine Locking implementations for tbot """
+
+import contextlib
+import typing
+import time
+import tbot
+from tbot.machine import machine, linux
+
+
+class MachineLock(machine.PreConnectInitializer):
+    """
+    Machine locking implementation based on Python and bash.
+    """
+
+    # lock expiry interval in seconds, 0 or less than lock_sleep does not expire
+    lock_expiry: int = 0
+    # number of locks in the pool
+    lock_pool_max: int = 1
+    # time to sleep in seconds for things to settle
+    lock_sleep: int = 1
+    # the number of this lock
+    lock_no: None
+    # the lock file used fo this lock
+    lock_file: None
+
+    @property
+    def lock_name(self) -> str:
+        """
+        Prefix from which lock file name is derived
+        """
+        return self.name
+
+    def _make_temp(self, labhost) -> linux.Path:
+        retval, result = labhost.exec("mktemp", "lock.XXX")
+        if retval == 0:
+            tempfile = linux.Path(labhost, result.strip())
+            if self.lock_expiry < self.lock_sleep:  # if lock does not expire
+                labhost.exec("echo", "0", linux.RedirStdout(tempfile))
+            else:  # else write the Unix time timestamp of expiry
+                labhost.exec(
+                    "echo",
+                    str(int(time.time()) + self.lock_expiry),
+                    linux.RedirStdout(tempfile),
+                )
+            # Append the PID of this context's shell
+            labhost.exec("echo", linux.Raw("$$"), linux.Raw(">>"), tempfile)
+            return tempfile
+        else:
+            raise Exception("Could not create tempfile")
+
+    def _locker(self, labhost, lockfile) -> bool:
+        tempfile = self._make_temp(labhost)
+        success = False
+        repeat = True
+        while repeat:  # do-while
```

Comment:
I'm not sure I understand what this loop is for.  We only attempt acquiring the lock once, right?  Or am I missing something here?

Reply:
The loop is there for when the lock was stale and deleted, so that we can try locking it again. Or for when we got the lock, but it someone else deleted it since we took it (as explained above).